go/host/db/rollups.go

```diff
@@ -0,0 +1,278 @@
+package db
+
+import (
+	"bytes"
+	"math/big"
+
+	gethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/obscuronet/go-obscuro/go/common"
+	"github.com/obscuronet/go-obscuro/go/common/log"
+)
+
+// TODO - #718 - Remove any rollup-based getters that are superseded by the batch-based getters.
+
+// DB methods relating to rollups.
+
+// headerKey = rollupHeaderPrefix  + hash
+func rollupHeaderKey(hash gethcommon.Hash) []byte {
+	return append(rollupHeaderPrefix, hash.Bytes()...)
+}
+
+// headerKey = rollupTxHashesPrefix + rollup hash
+func rollupTxHashesKey(hash gethcommon.Hash) []byte {
+	return append(rollupTxHashesPrefix, hash.Bytes()...)
+}
+
+// headerKey = rollupHashPrefix + number
+func rollupHashKey(num *big.Int) []byte {
+	return append(rollupHashPrefix, []byte(num.String())...)
+}
+
+// headerKey = rollupNumberPrefix + hash
+func rollupNumberKey(txHash gethcommon.Hash) []byte {
+	return append(rollupNumberPrefix, txHash.Bytes()...)
+}
+
+// GetHeadRollupHeader returns the header of the node's current head rollup, or (nil, false) if no such header is found.
+func (db *DB) GetHeadRollupHeader() (*common.Header, bool) {
+	headRollupHash, found := db.readHeadRollupHash(db.kvStore)
+	if !found {
+		return nil, false
+	}
+	return db.readRollupHeader(db.kvStore, *headRollupHash)
+}
+
+// GetRollupHeader returns the rollup header given the Hash, or (nil, false) if no such header is found.
+func (db *DB) GetRollupHeader(hash gethcommon.Hash) (*common.Header, bool) {
+	return db.readRollupHeader(db.kvStore, hash)
+}
+
+// AddRollupHeader adds a rollup's header to the known headers
+func (db *DB) AddRollupHeader(header *common.Header, txHashes []common.TxHash) {
+	b := db.kvStore.NewBatch()
+	db.writeRollupHeader(b, header)
```

Comment:
If one write fails should the other rollback ?

Reply:
I think so. Do you think we shouldn't?

Reply:
I think we should. But it seems like at the moment if one write fails the approach is to panic. Is that intended yeah ?

Reply:
Yeah - you want all the writes to succeed or fail together.

Reply:
I understand that's implicit, but what if there was a TODO mentioning that ? At the moment if one write fails the whole node stops, it's not an explicit logic of what its expected to happen.

Reply:
That'll be picked up by the follow-up ticket, if that's ok